03_letters/nicholascross/main_test.go

```diff
@@ -0,0 +1,79 @@
+package main
+
+import (
+	"bytes"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+)
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+
+	main()
+
+	expected := strconv.Quote(`a:2
+b:1
+`)
+	actual := strconv.Quote(buf.String())
+
+	if expected != actual {
+		t.Errorf("Unexpected output in main() %v", actual)
+	}
+}
+
+func TestSortLetters(t *testing.T) {
+	var testCases = map[string]struct {
+		input map[rune]int
+		want  []string
+	}{
+		"correct counts": {
+			input: map[rune]int{'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5},
+			want:  []string{"a:1", "b:2", "c:3", "d:4", "e:5"},
+		},
+		"sorted counts": {
+			input: map[rune]int{'c': 3, 'd': 4, 'e': 5, 'a': 1, 'b': 2},
+			want:  []string{"a:1", "b:2", "c:3", "d:4", "e:5"},
+		},
+		"empty": {input: map[rune]int{}, want: []string{}},
+	}
+
+	for name, test := range testCases {
+		input := test.input
+		want := test.want
+		t.Run(name, func(t *testing.T) {
+			assert.Equalf(t, sortLetters(input), want, "%v was not sorted", input)
+		})
+	}
+}
+
+func TestLetters(t *testing.T) {
+	var testCases = map[string]struct {
+		input string
+		want  map[rune]int
+	}{
+		"correct counts": {
+			input: "abbcccddddeeeee",
+			want:  map[rune]int{'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5},
+		},
+		"unordered counts": {
+			input: "ebdbcaddecdecee",
+			want:  map[rune]int{'a': 1, 'b': 2, 'c': 3, 'd': 4, 'e': 5},
+		},
+		"special characters": {
+			input: "无马e😊e马👍👍a👌😒",
+			want:  map[rune]int{'a': 1, '马': 2, '👍': 2, '👌': 1, 'e': 2, '无': 1, '😊': 1, '😒': 1},
+		},
+		"empty": {input: "", want: map[rune]int{}},
+	}
+
+	for name, test := range testCases {
+		input := test.input
+		want := test.want
```

Comment:
I prefer `expected` as a name over `want`.

Reply:
Sure, I received some early feedback that my variable names were too long, I actually prefer that too, I am still trying to find the right balance between brevity and clarity in `go`.

I will update 👍

Reply:
Yeah that makes sense, within tests I would generally always follow the `input / expected` convention. But for the same short lived variable within a normal function, I would be more concise with my naming. It is a bit of personal preference as well I guess.

Reply:
So I have just taken a look at the example we provide in the course notes, and it uses `want`. Hahaha maybe I am just as confused. Feel free to keep it as is as we are endorsing this in the notes 👍 Sorry if I am just causing further confusion.